File: src/leetcode/allnodesdistancekinbinarytree/AllNodesDistanceKInBinaryTree.java
package leetcode.allnodesdistancekinbinarytree;

import trees.TreeNode;

import java.util.*;
import java.util.stream.Collectors;

public class AllNodesDistanceKInBinaryTree {

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {

        Set<Integer> result = new HashSet<>();

        // Turn the tree into a graph
        Map<TreeNode, Set<TreeNode>> graph = traverse(root);
        HashSet<Object> V = new HashSet<>();
        V.addAll(graph.keySet());
        for(Set<TreeNode> adjacent : graph.values()){
            V.addAll(adjacent);
        }
        boolean[] visited = new boolean[501];
        dfs(graph, target, 0, k, visited, result);

        return result.stream().collect(Collectors.toList());

    }

    private void dfs(Map<TreeNode, Set<TreeNode>> graph,
                     TreeNode root,
                     int distance,
                     int k,
                     boolean[] visited,
                     Set<Integer> result) {
        visited[root.val] = true;

        if (distance == k) {
            result.add(root.val);
            return;
        }

        Set<TreeNode> adjacent = graph.get(root);
        if (adjacent != null && !adjacent.isEmpty()) {
            for (TreeNode v : adjacent) {
                if (!visited[v.val])
                    dfs(graph, v, distance + 1, k, visited, result);
            }
        }
    }

    private Map<TreeNode, Set<TreeNode>> traverse(TreeNode root) {
        Map<TreeNode, Set<TreeNode>> graph = new HashMap<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode polled = q.poll();
                if (polled.left != null) {
                    addEdge(graph, polled, polled.left);
                    addEdge(graph, polled.left, polled);
                    q.add(polled.left);
                }

                if (polled.right != null) {
                    addEdge(graph, polled, polled.right);
                    addEdge(graph, polled.right, polled);
                    q.add(polled.right);
                }
            }
        }

        return graph;
    }

    private void addEdge(Map<TreeNode, Set<TreeNode>> graph, TreeNode from, TreeNode to) {
        if (graph.containsKey(from)) {
            graph.get(from).add(to);
        } else {
            Set<TreeNode> adjacent = new HashSet<>();
            adjacent.add(to);
            graph.put(from, adjacent);
        }
    }


    public List<Integer> deadEnd(TreeNode root, TreeNode target, int k) {
        Queue<TreeNode> q = new LinkedList<>();
        Set<Integer> result = new HashSet<>();
        Map<Integer, Set<TreeNode>> nodesByLevel = new HashMap<>();
        q.add(root);

        int depth = 0;
        int targetLevel = 0;
        while (!q.isEmpty()) {
            Set<TreeNode> level = new HashSet<>();
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode polled = q.poll();
                level.add(polled);
                if (polled.right != null) q.add(polled.right);
                if (polled.left != null) q.add(polled.left);
            }

            nodesByLevel.put(depth, level);
            if (level.contains(target))
                targetLevel = depth;
            depth++;
        }

        Set<TreeNode> leftChildren = new HashSet<>();
        collect(root.left, leftChildren);
        Set<TreeNode> rightChildren = new HashSet<>();
        collect(root.right, rightChildren);

        Set<TreeNode> targetSubtree = leftChildren.contains(target) ? leftChildren : rightChildren;

        for (Map.Entry<Integer, Set<TreeNode>> levelNodes : nodesByLevel.entrySet()) {
            int level = levelNodes.getKey();
            Set<TreeNode> nodes = levelNodes.getValue();
            // targetLevel + level if the belong to different sub trees
            // targetLevel - level == k if they belong to the same subtree
            for (TreeNode node : nodes) {
                if ((targetSubtree.contains(node) && targetLevel - level == k) ||
                        (targetSubtree.contains(node) && level - targetLevel == k) ||
                        (targetSubtree.contains(node) && level + targetLevel == k && node != target) ||
                        (!targetSubtree.contains(node) && targetLevel + level == k))
                    result.add(node.val);
            }
        }

//        if (nodesByLevel.containsKey(targetLevel + k))
//            result.addAll(nodesByLevel.get(targetLevel + k).stream().map(node -> node.val).collect(Collectors.toList()));

        return new ArrayList<>(result);
    }

    private void collect(TreeNode root, Set<TreeNode> nodes) {
        if (root != null) {
            collect(root.left, nodes);
            nodes.add(root);
            collect(root.right, nodes);
        }
    }
}

File: python-code/all_nodes_distance_k_in_binary_tree.py
from tree_node import TreeNode
from typing import *

graph: Dict = {}


def find_target(root: TreeNode, target: int) -> TreeNode:
    if root.val == target:
        return root

    result_left = find_target(root.left, target)
    result_right = find_target(root.right, target)

    if result_left:
        return result_left

    return result_right


def create_graph(root: TreeNode):
    if root is None:
        return

    root_neighbors = graph.get(root, [])

    if root.left:
        left_child_neighbors = graph.get(root.left, [])
        left_child_neighbors.append(root)
        graph[root.left] = left_child_neighbors
        root_neighbors.append(root.left)
        create_graph(root.left)

    if root.right:
        right_child_neighbors = graph.get(root.right, [])
        right_child_neighbors.append(root)
        graph[root.right] = right_child_neighbors
        root_neighbors.append(root.right)
        create_graph(root.right)

    graph[root] = root_neighbors


def traverse_bfs(root: TreeNode,
                 distance: int,
                 k: int,
                 result: List[TreeNode]) -> None:
    if root is None:
        return

    if distance == k:
        result.append(root)

    traverse_bfs(root.left, distance + 1, k, result)
    traverse_bfs(root.right, distance + 1, k, result)


def tree_dfs(root: TreeNode,
             distance: int,
             k: int,
             visited: Set,
             result: List[TreeNode]):
    if root is None:
        return

    visited.add(root)

    if distance == k:
        result.append(root)

    for v in graph[root]:
        if v not in visited:
            tree_dfs(v, distance + 1, k, visited, result)


if __name__ == '__main__':
    visited: Set = set()
    result: List[TreeNode] = []
    a: TreeNode = TreeNode(1)
    a.left = TreeNode(2)
    a.right = TreeNode(3)
    # traverse_bfs(a, 0, 1, result)
    # while result:
    #     print(result.pop())

    create_graph(a)
    tree_dfs(a.left, 0, 2, visited, result)

    print(graph)
    print(result)
